Improve documentation for annotating generator functions (#11623)
python/mypy: docs/source/kinds_of_types.rst

Generators
**********

A basic generator that only yields values can be succinctly annotated as having a return
type of either :py:class:`Iterator[YieldType] <typing.Iterator>` or :py:class:`Iterable[YieldType] <typing.Iterable>`. For example:

.. code-block:: python

   def squares(n: int) -> Iterator[int]:
       for i in range(n):
           yield i * i
           

> In general, :py:class:`Iterator[YieldType] <typing.Iterator>` should be preferred over
> :py:class:`Iterable[YieldType] <typing.Iterable>` for the return type of a generator function,
> as it is more precise regarding the type of the returned object. All iterator objects have a
> :py:meth:`~iterator.__next__` method, so using ``Iterator`` as the return type alerts mypy to
> the fact that :py:func:`next` can be called on the object returned from the generator function.
> Using ``Iterable`` runs the risk of mypy raising false-positive errors if a user attempts to call
> `next()` on the returned object.
> 
> However, in some cases, it may be considered an implementation detail that ``next()`` can be
> called on the object returned from the generator function. In these cases, it may be preferable
> to use ``Iterable`` instead of ``Iterator``.

Reviewer: Nit rewording. My goals are a) trying to keep things brief and the recommendation clear, since new users' eyes tend to glaze over, b) calling out the rule of thumb explicitly
```suggestion
A good rule of thumb is to annotate functions with the most specific return type
that doesn't leak incidental implementation detail. In keeping with this, prefer
:py:class:`Iterator[YieldType] <typing.Iterator>` over
:py:class:`Iterable[YieldType] <typing.Iterable>`. In particular, using
``Iterator`` lets mypy know users are allowed to call :py:func:`next` on the
object returned from the generator function.
```

Author: My wording was too verbose, but I found "A good rule of thumb is to annotate functions with the most specific return type
that doesn't leak incidental implementation detail" to be too dense. What do you think of the latest iteration? 🙂

If you want your generator to accept values via the :py:meth:`~generator.send` method or return

> a value, you should use the

Reviewer: ```suggestion
a value, however, you should use the
```

Author: Thanks 😊

> a value, meanwhile, you should use the

Reviewer: ```suggestion
a value, you should use the
```

Author: Sorry to be a pain, but I actually quite liked the word "meanwhile" here! I think it creates a better flow between the paragraphs, more clearly indicating that *this* situation is in contrast to the *previous* situation discussed in the previous paragraph.

I'd be open to other words such as "however" 🙂

Reviewer: Okay, makes sense, I put in "however"!
(to explain my original nit here: I didn't originally read that "meanwhile" as referring to a previous paragraph, and without that context "meanwhile" connotes a kind of temporariness or incompleteness)

Author: I think the reason I went for "meanwhile" over "however" was that I'd used "however" just two sentences prior, fwiw 🙂 (I'm happy with "however", so don't want to belabor the point!!)

Generators
**********

A basic generator that only yields values can be annotated as having a return

> type of either :py:class:`Iterator[YieldType] <typing.Iterator>` or :py:class:`Iterable[YieldType] <typing.Iterable>`. For example:

Reviewer 2: I don't agree on this change. Sometimes you can refactor generators into regular function and vice a versa. Imagine this:

```python
def return_values() -> Iterable[int]:
    return [1, 2, 3]
```

Then you refactor it to be a generator instead:

```python
def return_values() -> Iterable[int]:
    yield 1
    yield 2
    yield 3
```

But, the return type can stay the same. 

Or, sometimes generators are not required and measuring performance says that it is faster / cheaper to just use a `list`. So, you do the opposite refactoring.

I often use this as a good practice to annotate functions like this as `Iterable`. I know that they might change. And I don't want to have a really specific return type.

Or, another case: you can design your functions to forbid using `next()` explicitly. It might matter for some generators.

Reviewer 2: Instead I propose to make it more clear: which return type do you need? `Iterable`, `Iterator`, or `Generator`. A table maybe? 🤔

Author: I think it's good for mypy to advocate, in general, using as specific a return-type as possible 🙂

If you're writing a function that is to be part of the public API for a library, it's important for other users of your code to know that `next()` can safely be called on the object that's returned from this function. If you refactor `return_values` from a regular function to a generator, I'd argue you *should* change the return-type annotation, as otherwise mypy will raise false-positive errors if a user of your hypothetical library tries to call `next()` on the returned object.

I do see your point, though — I'll have a think about ways I could change this proposed alteration to make it clearer 🙂
